src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionCache.java

```diff
@@ -53,23 +58,92 @@ public RemoteExecutionCache(
    * <p>Note that this method is only required for remote execution, not for caching itself.
    * However, remote execution uses a cache to store input files, and that may be a separate
    * end-point from the executor itself, so the functionality lives here.
+   *
+   * <p>Callers should pass {@code true} for {@code checkAll} on retries as a precaution to avoid
+   * getting stale results from other threads.
    */
   public void ensureInputsPresent(
       RemoteActionExecutionContext context,
       MerkleTree merkleTree,
-      Map<Digest, Message> additionalInputs)
+      Map<Digest, Message> additionalInputs,
+      boolean checkAll)
       throws IOException, InterruptedException {
     Iterable<Digest> allDigests =
         Iterables.concat(merkleTree.getAllDigests(), additionalInputs.keySet());
-    ImmutableSet<Digest> missingDigests =
-        getFromFuture(cacheProtocol.findMissingDigests(context, allDigests));
+    List<ListenableFuture<Void>> futures = new ArrayList<>();
+
+    if (checkAll || !options.experimentalRemoteDeduplicateUploads) {
+      ImmutableSet<Digest> missingDigests =
+          getFromFuture(cacheProtocol.findMissingDigests(context, allDigests));
+      for (Digest missingDigest : missingDigests) {
+        futures.add(uploadBlob(context, missingDigest, merkleTree, additionalInputs));
+      }
+    } else {
+      // This code deduplicates findMissingDigests calls as well as file uploads. It works by
```

Comment:
Why do we need to distinguish owned and unowned futures?

I think we are doing too much here. Can we have a cache layer between `RemoteExecutionCache` and `RemoteCacheClient` which is used to deduplicate the uploads and owns the futures. All the futures are unowned from the perspective of `RemoteExecutionCache`. e.g.

```
class UploadCache {
    ListenableFuture<Void> upload(MerkleTree merkleTree, Map<Digest, Message> additionalInputs);
}
```

We can also hide the `findMissingDigests` call inside `UploadCache` and can potentially save the call if the requested uploads are already uploaded or are in progress.

[AsyncTaskCache](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/util/AsyncTaskCache.java;l=50;bpv=1;bpt=1?q=Async&ss=bazel%2Fbazel) which is added recently to fix #12972 can be used to implement the cache.

FYI, we have the deduplication logic at [ByteStreamUploader](https://cs.opensource.google/bazel/bazel/+/refs/tags/4.0.0:src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java;l=209?q=ByteStreamUploader&ss=bazel%2Fbazel) but isn't enabled since the `forceUpload` is set to `true` [here](https://cs.opensource.google/bazel/bazel/+/refs/tags/4.0.0:src/main/java/com/google/devtools/build/lib/remote/GrpcCacheClient.java;l=400). That being said, the deduplication is only available for gRPC cache client. This PR allows deduplication for all cache protocols.

Reply:
> FYI, we have the deduplication logic at [ByteStreamUploader](https://cs.opensource.google/bazel/bazel/+/refs/tags/4.0.0:src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java;l=209?q=ByteStreamUploader&ss=bazel%2Fbazel) but isn't enabled since the `forceUpload` is set to `true` [here](https://cs.opensource.google/bazel/bazel/+/refs/tags/4.0.0:src/main/java/com/google/devtools/build/lib/remote/GrpcCacheClient.java;l=400). ~That being said, the deduplication is only available for gRPC cache client. This PR allows deduplication for all cache protocols.~ (This is only available for remote execution which only supports gRPC protocol)

I’m prototyping remote execution support also with HTTP cache protocol. Because that allows using HTTP load balance infrastructure. I get performance on par with gRPC. Therefore **I very much appreciate that this PR is generic and works also for HTTP**.

Reply:
I don't think AsyncTaskCache would be ideal here. The problem is that the `findMissingBlobs` calls are *bulk* calls: we have to deduplicate the individual entries, but we still want to perform bulk calls for performance. That requires passing a collection in and out of this class.

I think ByteStreamUploader is not a good place for a cache. The API to check for CAS entries is `findMissingBlobs`, which is a bulk API. ByteStreamUploader has a single file upload API - it can't really automatically turn a sequence of single-file uploads into a bulk check call. The caching has to be outside of ByteStreamUploader and itself needs a bulk API to be able to efficiently call `findMissingBlobs`. I also don't like that it unconditionally caches knowledge about an upload for the duration of the build - if the CAS loses entries faster than the build time, Bazel gets confused and fails for no good reason.

The distinction between owned and unowned futures is thus: there is any number of concurrent calls into this method, and we own some of the digests, but not all. We then make a bulk `findMissingBlobs` call for all owned digests, and then upload the ones that are missing remotely.

I'm happy to remove the flag and always enable this code.

Reply:
Thanks for the clarification. I am now agree with you that ByteStreamUploader can not be used to deduplicate `findMissingBlobs` calls.

**Feel free to ignore**: AsyncTaskCache can be used to deduplicate uploads just like what you achieved with ConcurrentHashMap combined with deduplication logic.

> I'm happy to remove the flag and always enable this code.

Please remove the flag!

Can you please also add test cases for the deduplication?